### PR TITLE
WIP/RFC: interpolate: use a single source for pythran-compiled and alternative backends 

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

x-ref https://github.com/scipy/scipy/pull/23746, https://github.com/scipy/scipy/pull/23447

#### What does this implement/fix?

https://github.com/scipy/scipy/pull/23746 added a "generic" Array API support to RBFInterpolator, giving a 2x--15x perf boost on GPUs.
The price to pay is having to maintain two separate "backends", one for pythran-compiled numpy array inputs, and another one for alternative array libraries.

Here is an experiment to have a single source for
- a pythran AOT compiled extension
- alternative Array API arrays in the eager mode
- potentially, JIT compilation with alternative backends (jax.jit, torch.compile)

This is experimental and is not mergeable yet: we use an in-development pythran feature, from https://github.com/serge-sans-paille/pythran/pull/2371
Thus posting the PR early for viz and feedback.
 
TODO:
- [ ] check performance. https://github.com/serge-sans-paille/pythran/pull/2371#issuecomment-3708133157 indicates there is a performance hit for the "classic" numpy backend
- [ ] check compatibility with jitting, cf https://github.com/scipy/scipy/pull/23447

#### Additional information

There are two key features here:

1. A new pythran annotation syntax for the `xp` argument. Basically, it enables the following:

```
# pythran export foo(float[:], numpy pkg)    # new "numpy pkg" annotation
def foo(x, xp):                                                     # our typical "xp" argument
     return xp.sum(x, axis=-1)
```

Previously, the problem was that while  "generic" functions have to receive the `xp` argument, pythran was not allowing  namespace/modules as arguments.

With this new syntax: 
- At build time, we install both pythran-compiled extension _and its source_.
- At runtime, we use the compiled extension for numpy inputs (and thus pythran only sees  `numpy` as its `xp` argument) and the pure python source for non-numpy arrays. 

Since pythran annotations are just code comments, the source is valid python, and alternative array backends do not know or care about pythran. The reverse is also true: pythran knows nothing about non-numpy inputs.

The branching at runtime we still need to do manually. This PR does it with `_rbfinterp_xp.py` and `_rbfinterp_np.py` trampoline modules, just the same way as in main. The main difference is that we significantly cut on code duplication between pythran and "generic" backends.

2. We need to rewrite the pythran sources in a "vectorized" style. Previously, it was using the "C style" with explicit looping over arrays:

```
out = np.empty(n, dtype=float)
for i in range(n):
    out[i] = ...
```

This does not work well with non-numpy arrays, in either eager or jit modes, so we have to use a "vectorized" implementation style everywhere. This puts pressure on the compiler to inline temporaries etc, and currently  it looks like getting to perf parity might require some more work.

### CODE REVIEW

**Comment by ev-br** [comment]
CI failures of this kind

```
CRITICAL I am in trouble. Your input file does not seem to match Pythran's constraints...
../scipy/interpolate/_rbfinterp_pythran_src.py:4:29 error: Unsupported identifier `numpy` at that point.
----
- AOT compiled by pythran (hence # pythran export comments)
                             ^~~~ (o_0)
```
simply mean that that particular CI job installs a released pythran not a branch. Since the whole "install from a branch" thing is temporary, I think it's enough for demo purposes to not bother about all of our CI.  As long as there is at least one  with-pythran job and at least one no-pythran job which pass.

**Comment by ilayn** [comment]
Not sure how much we are benefiting from pythran but we can quickly make those as low level code since I don't see pythran/cython code growing and we are writing more C/C++ than ever lately. 

In fact I don't remember when we added significant pythran and/or cython combo. Not saying we shouldn't but if it is getting more in the way than it is helping maybe we can get those parts out to C++ and move on instead of carrying the weight around all the time

**Comment by ev-br** [comment]
Yes, we definitely can rewrite RBFInterpolator in C. However:

- some people do prefer `pythran` to manual C [1].

- in one word: GPUs. 
    Speedups of 15x are right on the table, additional small effort promises 40x [2].
    SciPy definitely cannot carry its own CUDA C kernels, so to benefit we need to maintain two versions, one in C and one 
    via Array API. What pythran promises is a single source for the two.

[1] not you or me, but these people certainly do exist.
[2] e.g. https://github.com/ev-br/bench_playground/blob/master/viz.rbf.ipynb, scroll down to cell [23]

**Comment by ilayn** [comment]
What I meant is that we can remove the pythran code and move to Cython, Python or C/C++. As a dependency we have very little code that runs pythran and I don't think it justifies its existence as a dependency so far.

> What pythran promises is a single source for the two.

Well, ain't that a promise. Isn't that what we are doing everywhere else? Not sure I buy that argument to be honest. No idea what I am looking at in the plots (especially in the laptop section). I thought the point of the work is to keep code in python so that array vendors can work natively.

**Comment by ev-br** [comment]
> Well, ain't that a promise. Isn't that what we are doing everywhere else?

Yes. Elsewhere we do it for pure python numpy-only code, here we do it for compiled code.

IOW, consider a silly example:

```
x = xp.arange(5)
y = xp.sin(x)
```

When `xp is numpy`, the second line launches a CPU kernel for elementwise sine function.
When `xp is torch`, it launches a CUDA kernel.
Actually, it launches a CPU or a CUDA kernel, depending on which device `x` arrays is. Either way, both kernels live in pytorch itself, scipy is just a user.

Now, this works when assembling kernels from somewhere else (numpy, torch, you name it). What to do when scipy itself needs a compiled kernel on CPU?
This PR is a demonstration of a possible answer that `pythran` master branch allows.

>  the point of the work is to keep code in python so that array vendors can work natively.

Exactly. On numpy, we keep our pythran-compiled performance;  with other array libraries, they work their magic. For us, we only keep a single python source with annotations in `# pythran export` code comments. 

> No idea what I am looking at in the plots (especially in the laptop section)

Can ignore the "laptop" section if it gets in the way.
The cell 23 plot I was referring to is this one:
 
image

The horizontal axis is the "problem size" and the vertical axis is the speedup relative to scipy main. So for sizes > 200,  pytorch eager CUDA (i.e. just pure python xp-using code, no shenanigans) is ~25x faster than pythran-compiled CPU version.

**Comment by ilayn** [comment]
Yes this much I understand but I don't understand why Pythran is so slow compared to eager pytorch. These are not exceptionally big problems so there should not be this much of difference. Hence my comment; if Pythran is not really giving us exceptional performance then why do we keep Pythran around as a dependency? Because we really have 5 6 code snippets for it across SciPy.

We can just swap it out with something else whatever that might be and have one less thing to worry about. I'm not saying Cython would be any better. But maybe the way we write the Python code can be improved with a C++ code since not all users are running code on GPU let alone NVIDIA ones (if you can buy one that is these days).

**Comment by ev-br** [comment]
> why Pythran is so slow compared to eager pytorch. 

Not sure what is this based on. Do you refer to single-core CPU pythran example being just 25x slower than pytorch GPU eager, or that eager pytorch 4-core CPU being only 5x slower than single-core pythran on the same CPU (cell 13), or something else entirely?

We could rewrite our pythran kernels into plain C and drop the dependency, sure. Should we though? Ain't broken, I'd think. So I'd explore the current OP option first. If it gets us a bunch of benefits from torch/jax for free, that is.
